keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,430 @@
+* NoRestart - the default value; resize the Container without restarting it,
+* RestartContainer - restart the Container in-place to apply new resource
+  values (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problems whereas
+changes to available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### Pod Resize Retry Policy
+
+If resource update fails, say due to lack of space on Node, default behavior
+is to let the initiating actor such as VPA handle the failure. Alternately, a
+Controller can either retry the update, or reschedule the Pod based on policy.
+
+PodSpec is extended with a new flag, PodSpec.RetryPolicy, with possible values:
+* NoRetry - the default value; do nothing, initiating actor handles failure,
+* RetryUpdate - Controller retries resource update in-place,
+* Reschedule - Controller evicts Pod, and creates updated Pod for scheduling.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
+
+Presently, when a Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod. In order to allow Pod resizing operation to work
+well with new Pod scheduling workflow, this design keeps Scheduler in the Pod
+resizing decision path.
+
+The following steps denote a typical flow of an in-place resize operation for a Pod
+with ResizePolicy set to Update for all its Containers.
+
+1. The initiating actor updates ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements
+   (e.g. limits are not below requested resources, QoS class does not change).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers reject the update, they set PodResizeSuccess
+     PodCondition's Reason to Rejected<Reason>, and set Status to false.
+1. API Server updates PodSpec object with the new desired ResourceRequirements.
+1. Scheduler observes that ResourceRequirements and ResourceBindings differ.
```

Comment:
i am concerned about backwards compatibility.  there are many schedulers in the world that schedule pods.  historically, the daemonset controller acted as its own scheduler, and would have had to also follow this flow.  i think we need to assume that operators may self-schedule pods.  this would potentially conflict with the flow proposed.

/cc @liggitt - what are your thoughts here?

Reply:
When you say "self-scheduling", are you concerned that users may be hard-coding the pod's nodeName, and thus bypassing this step?  Or are you concerned that other scheduling actors would need to adopt the same logic of setting ResourceBindings to continue working?  If the latter is your concern, what do you think of the idea of building this logic into the `Binding` subresource?  I.E. Binding pod A to node N not only sets the `nodeName` of pod A, but also sets `ResourceBindings` to the pod's `ResourceRequirements`.  Then all schedulers continue to function.

Reply:
i was concerned about controllers not authored from core k8s that may have acted as their own scheduler.  i was also concerned about projects that acted as kubelets and may not have a clear signal when a pod is ready to be executed (for example virtual kubelet).  if we are saying bind would bind both node and resources, its possible that would work, we still need to update the kubelet though to say "i can only run a pod if both the pod and its resources are bound".  in general, i still dislike the scheduler having to act as an intermediary to resource resizing for an already scheduled pod.  i think that still should be a dialogue between kubelet and api server where kubelet attempts to converge.

Reply:
Scheduler is not in the loop now, resolving this issue per updated design.